src/posix/platform/netif.cpp

```diff
@@ -379,6 +793,284 @@ static void processNetifLinkEvent(otInstance *aInstance, struct nlmsghdr *aNetli
+    size_t              buffer_len = rtm->rtm_msglen;
+
+    addr6.sin6_family   = 0;
+    netmask.sin6_family = 0;
+
+    if ((rtm->rtm_type == RTM_NEWADDR) || (rtm->rtm_type == RTM_DELADDR))
+    {
+        ifam = reinterpret_cast<struct ifa_msghdr *>(rtm);
+
+        VerifyOrExit(ifam->ifam_index == static_cast<unsigned int>(sTunIndex), OT_NOOP);
+
+        addrbuf  = (uint8_t *)&ifam[1];
+        addrmask = (unsigned int)ifam->ifam_addrs;
+    }
+#ifdef RTM_NEWMADDR
+    else if ((rtm->rtm_type == RTM_NEWMADDR) || (rtm->rtm_type == RTM_DELMADDR))
+    {
+        ifmam = reinterpret_cast<struct ifma_msghdr *>(rtm);
+
+        VerifyOrExit(ifmam->ifmam_index == static_cast<unsigned int>(sTunIndex), OT_NOOP);
+
+        addrbuf  = (uint8_t *)&ifmam[1];
+        addrmask = (unsigned int)ifmam->ifmam_addrs;
+    }
+#endif
+    addrlen = (unsigned int)buffer_len;
+
+    if (addrmask != 0)
+    {
+        for (i = 0; i < RTAX_MAX; i++)
+        {
+            unsigned int mask = (addrmask & (1 << i));
+            if (mask)
+            {
+                sa = (struct sockaddr *)addrbuf;
+
+                if (sa->sa_family == AF_INET6)
+                {
+                    if (i == RTAX_IFA)
+                        memcpy(&addr6, sa, sizeof(sockaddr_in6));
+                    if (i == RTAX_NETMASK)
+                        memcpy(&netmask, sa, sizeof(sockaddr_in6));
+                }
+                addrlen -= SA_SIZE(sa);
+                addrbuf += SA_SIZE(sa);
+            }
+        }
+    }
+
+    if (addr6.sin6_family == AF_INET6)
+    {
+        is_link_local = false;
+        if (IN6_IS_ADDR_LINKLOCAL(&addr6.sin6_addr))
+        {
+            is_link_local = true;
+            // clear the scope -- Mac OS X sends this to us (bozos!)
+            addr6.sin6_addr.s6_addr[3] = 0;
+        }
+        else if (IN6_IS_ADDR_MC_LINKLOCAL(&addr6.sin6_addr))
+        {
+            addr6.sin6_addr.s6_addr[3] = 0;
+        }
+
+        ot::Ip6::Address addr;
+        memcpy(&addr, &addr6.sin6_addr, sizeof(addr));
+
+        if (rtm->rtm_type == RTM_NEWADDR
+#ifdef RTM_NEWMADDR
+            || rtm->rtm_type == RTM_NEWMADDR
+#endif
+        )
+        {
+            if (!addr.IsMulticast())
+            {
+                otNetifAddress netAddr;
+                bool           subscribed;
+
+                netAddr.mAddress      = addr;
+                netAddr.mPrefixLength = (uint8_t)calculate_prefix_len(&netmask);
+
+                subscribed = unicastAddressIsSubscribed(aInstance, &netAddr);
+
+                if (subscribed)
+                {
+                    inet_ntop(AF_INET6, addr6.sin6_addr.s6_addr, addressString, sizeof(addressString));
+                    otLogNotePlat("ADD [U] %s (already subscribed, ignored)", addressString);
+                }
+                else
+                {
+                    if (is_link_local)
+                    {
+                        // remove the stack-added link-local address
```

Comment:
Why shall we remove stack-added link local address from the interface address?

Reply:
Because it breaks things :-).

Stacks normally add their own link-local IPv6 address, though they may have options to avoid doing so -- but unfortunately, some stacks (mac OS in particular) don't.

When we get notified that the address has been added (as is happening here), openthread will not allow you to add this (second) link-local address to its list of addresses (I believe openthread rejects all addition of link-local addresses, as they are internally handled).  

So if we do NOT remove it, then the stack will have an address that openthread doesn't know about.

But worse than that, the stack will create packets with that address as a source address, and send them down to openthread for sending onto the network.  Other nodes will attempt to respond, but the openthread stack will not "receive" these packets because they're not bound for a known address.